## 任务一： 论文数据统计
###  1.1  任务说明  
- 任务主题：论文数量统计，即统计2019年全年计算机各个方向论文数量；
- 任务内容：赛题的理解、使用 Pandas 读取数据并进行统计；
- 任务成果：学习 Pandas 的基础操作；  

### 1.2 数据集介绍  
*数据集来源：* [数据集链接](https://www.kaggle.com/Cornell-University/arxiv)  
*数据集的格式如下：*  
- id：arXiv ID，可用于访问论文；
- submitter：论文提交者；
- authors：论文作者；
- title：论文标题；
- comments：论文页数和图表等其他信息；
- journal-ref：论文发表的期刊的信息；
- doi：数字对象标识符，https://www.doi.org；
- report-no：报告编号；
- categories：论文在 arXiv 系统的所属类别或标签；
- license：文章的许可证；
- abstract：论文摘要；
- versions：论文版本；
- authors_parsed：作者的信息。

### 1.3 数据集类别如下：
我们从arxiv官网，查询到论文的类别名称以及其解释如下。

链接：https://arxiv.org/help/api/user-manual 的 5.3 小节的 Subject Classifications 的部分，或 https://arxiv.org/category_taxonomy， 具体的153种paper的类别部分如下：  
'astro-ph': 'Astrophysics',
'astro-ph.CO': 'Cosmology and Nongalactic Astrophysics',
'astro-ph.EP': 'Earth and Planetary Astrophysics',
'astro-ph.GA': 'Astrophysics of Galaxies',
'cs.AI': 'Artificial Intelligence',
'cs.AR': 'Hardware Architecture',
'cs.CC': 'Computational Complexity',
'cs.CE': 'Computational Engineering, Finance, and Science',
'cs.CV': 'Computer Vision and Pattern Recognition',
'cs.CY': 'Computers and Society',
'cs.DB': 'Databases',
'cs.DC': 'Distributed, Parallel, and Cluster Computing',
'cs.DL': 'Digital Libraries',
'cs.NA': 'Numerical Analysis',
'cs.NE': 'Neural and Evolutionary Computing',
'cs.NI': 'Networking and Internet Architecture',
'cs.OH': 'Other Computer Science',
'cs.OS': 'Operating Systems',

### 1.4 数据概览

In [1]:
# 导入所需的package
import seaborn as sns #用于画图
from bs4 import BeautifulSoup #用于爬取arxiv的数据
import re #用于正则表达式，匹配字符串的模式
import requests #用于网络连接，发送网络请求，使用域名获取对应信息
import json #读取数据，我们的数据为json格式的
import pandas as pd #数据处理，数据分析
import matplotlib.pyplot as plt #画图工具

In [2]:
# 读入数据

data  = [] #初始化
#使用with语句优势：1.自动关闭文件句柄；2.自动显示（处理）文件读取数据异常
with open(r"D:/offer/Github/DataWhale/arxiv-metadata-oai-2019.json/arxiv-metadata-oai-2019.json", 'r') as f: 
    for line in f: 
        data.append(json.loads(line))
        
data = pd.DataFrame(data) #将list变为dataframe格式，方便使用pandas进行分析
data.shape #显示数据大小

(170618, 14)

In [3]:
data.head()

id             submitter  \
0  0704.0297        Sung-Chul Yoon   
1  0704.0342  Patrice Ntumba Pungu   
2  0704.0360          Zaqarashvili   
3  0704.0525        Sezgin Ayg\"un   
4  0704.0535        Antonio Pipino   

                                             authors  \
0  Sung-Chul Yoon, Philipp Podsiadlowski and Step...   
1                          B. Dugmore and PP. Ntumba   
2                   T.V. Zaqarashvili and K Murawski   
3          Sezgin Aygun, Ismail Tarhan, Husnu Baysal   
4  Antonio Pipino (1,3), Thomas H. Puzia (2,4), a...   

                                               title  \
0  Remnant evolution after a carbon-oxygen white ...   
1  Cofibrations in the Category of Frolicher Spac...   
2  Torsional oscillations of longitudinally inhom...   
3  On the Energy-Momentum Problem in Static Einst...   
4  The Formation of Globular Cluster Systems in M...   

                                            comments  \
0  15 pages, 15 figures, 3 tables, submitted to M...   
1                                           27 pages   
2                6 pages, 3 figures, accepted in A&A   
3  This submission has been withdrawn by arXiv ad...   
4  32 pages (referee format), 9 figures, ApJ acce...   

                      journal-ref                               doi report-no  \
0                            None  10.1111/j.1365-2966.2007.12161.x      None   
1                            None                              None      None   
2                            None        10.1051/0004-6361:20077246      None   
3  Chin.Phys.Lett.24:355-358,2007        10.1088/0256-307X/24/2/015      None   
4    Astrophys.J.665:295-305,2007                    10.1086/519546      None   

  categories license                                           abstract  \
0   astro-ph    None    We systematically explore the evolution of t...   
1    math.AT    None    Cofibrations are defined in the category of ...   
2   astro-ph    None    We explore the effect of an inhomogeneous ma...   
3      gr-qc    None    This paper has been removed by arXiv adminis...   
4   astro-ph    None    The most massive elliptical galaxies show a ...   

                                            versions update_date  \
0  [{'version': 'v1', 'created': 'Tue, 3 Apr 2007...  2019-08-19   
1  [{'version': 'v1', 'created': 'Tue, 3 Apr 2007...  2019-08-19   
2  [{'version': 'v1', 'created': 'Tue, 3 Apr 2007...  2019-08-19   
3  [{'version': 'v1', 'created': 'Wed, 4 Apr 2007...  2019-10-21   
4  [{'version': 'v1', 'created': 'Wed, 4 Apr 2007...  2019-08-19   

                                      authors_parsed  
0  [[Yoon, Sung-Chul, ], [Podsiadlowski, Philipp,...  
1                 [[Dugmore, B., ], [Ntumba, PP., ]]  
2         [[Zaqarashvili, T. V., ], [Murawski, K, ]]  
3  [[Aygun, Sezgin, ], [Tarhan, Ismail, ], [Baysa...  
4  [[Pipino, Antonio, ], [Puzia, Thomas H., ], [M...

In [4]:
data.columns

Index(['id', 'submitter', 'authors', 'title', 'comments', 'journal-ref', 'doi',
       'report-no', 'categories', 'license', 'abstract', 'versions',
       'update_date', 'authors_parsed'],
      dtype='object')

In [7]:
#先粗略统计论文类别信息
'''
count：一列数据的元素个数；
unique：一列数据中元素的种类；
top：一列数据中出现频率最高的元素；
freq：一列数据中出现频率最高的元素的个数；
'''
data['categories'].describe()

count     170618
unique     15592
top        cs.CV
freq        5559
Name: categories, dtype: object

以上的结果表明：共有170618个数据，有15592个子类（因为有论文的类别是多个，例如一篇paper的类别是CS.AI & CS.MM和一篇paper的类别是CS.AI & CS.OS属于不同的子类别，这里仅仅是粗略统计），其中最多的种类是cs.CV，，共出现了5559次。

In [8]:
#得到2019年以后的所有种类的论文
data["year"] = pd.to_datetime(data["update_date"]).dt.year #将update_date从例如2019-02-20的str变为datetime格式，并提取处year
del data["update_date"] #删除 update_date特征，其使命已完成
data = data[data["year"] >= 2019] #找出 year 中2019年以后的数据，并将其他数据删除
# data.groupby(['categories','year']) #以 categories 进行排序，如果同一个categories 相同则使用 year 特征进行排序
data.reset_index(drop=True, inplace=True) #重新编号
data #查看结果

id             submitter  \
0              0704.0297        Sung-Chul Yoon   
1              0704.0342  Patrice Ntumba Pungu   
2              0704.0360          Zaqarashvili   
3              0704.0525        Sezgin Ayg\"un   
4              0704.0535        Antonio Pipino   
...                  ...                   ...   
170613  quant-ph/9904032         Mikhail Lukin   
170614  solv-int/9511005            Wen-Xiu Ma   
170615  solv-int/9809008       Victor Enolskii   
170616  solv-int/9909010   Pierre van Moerbeke   
170617  solv-int/9909014         David Fairlie   

                                                  authors  \
0       Sung-Chul Yoon, Philipp Podsiadlowski and Step...   
1                               B. Dugmore and PP. Ntumba   
2                        T.V. Zaqarashvili and K Murawski   
3               Sezgin Aygun, Ismail Tarhan, Husnu Baysal   
4       Antonio Pipino (1,3), Thomas H. Puzia (2,4), a...   
...                                                   ...   
170613  V. A. Sautenkov, M. D. Lukin, C. J. Bednar, G....   
170614                     Wen-Xiu Ma, Benno Fuchssteiner   
170615  J C Eilbeck, V Z Enol'skii, V B Kuznetsov, D V...   
170616            M. Adler, T. Shiota and P. van Moerbeke   
170617                       D.B. Fairlie and A.N. Leznov   

                                                    title  \
0       Remnant evolution after a carbon-oxygen white ...   
1       Cofibrations in the Category of Frolicher Spac...   
2       Torsional oscillations of longitudinally inhom...   
3       On the Energy-Momentum Problem in Static Einst...   
4       The Formation of Globular Cluster Systems in M...   
...                                                   ...   
170613  Enhancement of Magneto-Optic Effects via Large...   
170614  Explicit and Exact Solutions to a Kolmogorov-P...   
170615  Linear r-Matrix Algebra for a Hierarchy of One...   
170616                                Pfaff tau-functions   
170617  The General Solution of the Complex Monge-Amp\...   

                                                 comments  \
0       15 pages, 15 figures, 3 tables, submitted to M...   
1                                                27 pages   
2                     6 pages, 3 figures, accepted in A&A   
3       This submission has been withdrawn by arXiv ad...   
4       32 pages (referee format), 9 figures, ApJ acce...   
...                                                   ...   
170613                                               None   
170614  14pages, Latex, to appear in Intern. J. Nonlin...   
170615                              plain LaTeX, 28 pages   
170616                                           42 pages   
170617                        13 pages, latex, no figures   

                           journal-ref                               doi  \
0                                 None  10.1111/j.1365-2966.2007.12161.x   
1                                 None                              None   
2                                 None        10.1051/0004-6361:20077246   
3       Chin.Phys.Lett.24:355-358,2007        10.1088/0256-307X/24/2/015   
4         Astrophys.J.665:295-305,2007                    10.1086/519546   
...                                ...                               ...   
170613                            None        10.1103/PhysRevA.62.023810   
170614                            None      10.1016/0020-7462(95)00064-X   
170615                            None                              None   
170616                            None                              None   
170617                            None       10.1088/0305-4470/33/25/307   

       report-no                                categories license  \
0           None                                  astro-ph    None   
1           None                                   math.AT    None   
2           None                                  astro-ph    None   
3           None             